Error Reporter Save name and email locally

DanielMurphy22 opened the pull request:
**Description of work.**

- User ContactInfo can now be saved and prefilled in the errorreporter to increase the number of reports with contact info (more convenient for users)
- It is now possible to send a textbox when selecting Share Non-Identifiable info. This also negates the need for fixing when to grey/ungrey out this button based on the textbox. This button is now only greyed out if the name or email are altered from empty or the saved+prefilled name and email.
- I tried to fix the tab ordering but QT seems to want to include other elements despite me setting their FocusPolicy to ClickFocus. I don't think this is important as with the initial focus being on the textbox, tabbing here just adds a tab to the text rather than jumping between elements.
- 
**To test:**

Either 1. setup a local error reporter server using the readme here: https://github.com/mantidproject/errorreports
(basically pull that repo, add the .env file, install docker-compose, run sudo bin/boot.sh and make sure to add `errorreports.rooturl = http://localhost:8082` to your Mantid.user.properties file)
or 2. be happy with spamming the ErrorReporter Slack channel for a bit!

- Make sure you have downloaded the TrainingCourseData
- In MantidWorkbench, Load `TrainingCourseData/Training_Exercise3a_SNS.nxs`
- Then open the algorithm dialog for NormaliseToMonitor
-  you should get an error reporter appear, but Mantid stays open, so you can open the error reporter many times quickly.

- Open the error reporter
- Input your name and email, tick the new SaveContactInfo checkbox and click `Share All Info`
- Open the error reporter: have the name and email prefilled and is the checkbox preticked?
- Try again with just the email and just the name. Notice how any changes get updated as long as the new checkbox is ticked
- Open the error reporter and Uncheck the tickbox
- Reopen and notice that the ContactInfo was deleted (the contact_info file was emptied)

- Also notice that you can type in the textbox and send non-identifiable info aslong as the name and email are not altered.

Fixes https://github.com/mantidproject/mantid/issues/28083

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
-->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

DanielMurphy22 committed:
Save Contact info in Error Reporter

Add checkbox for users to Save their name and email in a local file
They are only saved if the ShareAllInfo button is clicked
If the SaveContactInfo Checkbox is unclicked and any of the 3 options
are chosen then the Saved Contact info file is emptied
Saving only one of Name/Email works

DanielMurphy22 committed:
Minor updates

Allowed Sending FreeTextBox for ShareNonIdentifiableInfo
Tried to set tab order to Text/Name/Email

DanielMurphy22 committed:
Add release notes

cailafinn reviewed (commented):
Works really well, saves when checked, cleared when unchecked. Should really help with getting more consistent error reports. 
Just before approval, would it make sense to have the box checked by default? Might encourage more users to make use of it

DanielMurphy22 commented:
> Just before approval, would it make sense to have the box checked by default? Might encourage more users to make use of it

I had wondered this myself Conor and I was tempted to do just that. I thought that since the error reporter seems to be designed for user to opt-in when sharing details, I'd make this checkbox work on an opt-in basis. @martyngigg do you think we should have it this way or should it be checked by default?

martyngigg commented:
Ah this is really nice and will help us get better quality reports I think. Could I make a tiny suggestion that we change the label "Save Contact Info" to "Remember Me" as the latter phrase is a bit more widely used on forms that save details?

I would say that the default should still be off though. I think it's better to opt-in to saving information than accidentally having it saved and not realising.

DanielMurphy22 committed:
Rename Checkbox to RememberMe and ShareNonID corrections

Pass text_box to c++ to correctly send to server for Share NonIdentifiable
Ensured that Share NonIdentifiable only greys out
if name/email is empty or same as the prefilled value
Fix ErrorReporterTest.h

cailafinn reviewed (dismissed):
Looks good to me and all seems to work as expected. Hopefully will result in better reports from here on out!

martyngigg commented on qt/python/mantidqt/mantidqt/dialogs/errorreports/report.py:
```diff
@@ -79,6 +81,23 @@ def __init__(self, parent=None, show_continue_terminate=False):
         # Set default focus to the editing box, rather then letting qt try and guess
         self.input_free_text.setFocus()
 
+        self.contact_info_file_path = os.path.join(ConfigService.getAppDataDirectory(),
```
Could I suggest that rather than using a custom file and format that we us the `QSettings` functionality? Or possibly I've misunderstood why that couldn't be used?

DanielMurphy22 replied:
It's much cleaner with QSettings! I've not used them before. I think I've got it working. Do the self.CONTACT,NAME,EMAIL map to their save location?

DannyHindson committed:
Change line number on cpp check suppression

DanielMurphy22 committed:
Save contact info in QSettings

Update test_errorreport_presenter mock

DanielMurphy22 committed:
Merge branch 'main' into ER_Save_contact_info

cailafinn reviewed (changes requested):
Definitely much cleaner for the filesystem with the QSettings instead. Just noticed a tiny typo in the release notes

DanielMurphy22 committed:
Set QSettings defaults for ErrorReporter

QSettings save in the mantidworkbench.ini, even when
errorreporter launched separately after hard crash.
Also updated tooltips and release note.

DanielMurphy22 committed:
Merge branch 'main' into ER_Save_contact_info

DanielMurphy22 committed:
Duplicate Config strings

DanielMurphy22 committed:
Merge branch 'ER_Save_contact_info' of https://github.com/mantidproject/mantid into ER_Save_contact_info

martyngigg reviewed (changes requested):
Just one comment about an `if` that nearly tripped me up reading the code but then I think it is good.

DanielMurphy22 committed:
Remove unnecessary condition

martyngigg reviewed (approved):
This looks good!

A great UX improvement for users

cailafinn reviewed (approved):
Looking good here! Nice one 👍🏻
